### DISCUSSION TITLE
Portals

### DISCUSSION THREAD

**Discussion by eric-burel** [withastro/roadmap#1032] [Proposal]
# Summary

In addition to slots that work on parent-children relationship, introduce a concept of portals that works from any-level of nesting.

# Background & Motivation

The component hierarchy is directly tied to the DOM hierarchy which is not always the best structure for implementing a specific feature.

I am currently thinking of 2 use cases though some more might emerge during discussions:
- triggering dialogs: floating elements tend to have a specific location very high in the DOM to avoid being parasited by intermediate elements positioning logic, that's typically how you implement absolutely positioned dialogs with a guarantee that the position is computed based on the viewport and not some random parent element that used a CSS filter. 
- feeding the <head> section with metadata: for this use case, we might want to run further investigation to understand the impact of not having a meta tag into the `<head>` section. It's certainly not clean, but I suspect it does have an impact on SEO too.

For the first use case, at the moment you may use portals from your favorite client-side library, maybe this would just need some documentation to tell whether portals are supported or not.

However for proper SEO you may want the injection into the head tag to happen during server-side rendering only, it would be ideal to support this directly in Astro.

I've encountered the second use case while working on improving [TutorialKit metadata](https://github.com/stackblitz/tutorialkit/pull/342). 

Multiple changes were needed to make a proper connection between `.mdx` files and the `Layout` component that renders the meta tags:
- supporting a new fields in the .mdx content frontmatter
- crafting a YAML friendly representation of the metadata
- updating all types accordingly
- updating the layout component accordingly
- converting back the YAML representation into HTML

And we still only support a few metadata : image, title, description. A similar work would be needed for more advanced tags like alternate links. 

# Goals

- Bring a portal system

![image](https://github.com/user-attachments/assets/6ab9fc8e-10cb-4889-8cdd-8ea0a6c9a707)

# Example

```md
# lesson.mdx
title: My lesson
---
<portal to="head">
  <link rel="alternate" href="/foobar.com">
<portal />
```

```
// pages/[slug].astro
// assuming you get the content from a collection
const MarkdownContent = await lesson.render()
---
<head>
  <portal name="head />
</head>
<body>
  <!-- This component will also inject an alternate URL into the head-->
  <MarkdownContent />
</body>
```

Related:
https://github.com/stackblitz/tutorialkit/pull/342
https://github.com/withastro/astro/issues/8703
https://react.dev/reference/react-dom/createPortal

**Comment by Fryuni** [comment]
This seems completely doable from userland as an integration.

Enabling it (regardless if by an integration or built into Astro) would disable streaming responses for any page containing a portal tho, since it's contents can only be known when every component renders.

**Reply by Fryuni** [reply]
Implemented an integration for it: https://github.com/Fryuni/inox-tools/pull/167

Works with Astro components and with UI framework components (tho I still need to add the proper types for all of them, only React and Preact so far).

Once I add the types and tests I'll publish it

**Reply by eric-burel** [reply]
@Fryuni thanks for the reply, your code was super insightful!
My takeaways:
- You are implementing the portals in user-land by working on the rendered HTML, as a post-processing step. I deplore the lack of a similar feature in Next.js, which I would call an ["endware"](https://github.com/vercel/next.js/discussions/57042), because it unlocks so many user-land tweaks like this. It's also a possible approach for injecting polyfills depending on the user agent while keeping a page dynamic.
- It works server-side as expected, Astro middlewares run [at build-time only for static pages](https://docs.astro.build/en/guides/middleware/) so it's also static-friendly.
- A downside might be the cost, you need to parse the whole HTML content to find the portal. Ideally, this should be done against a kind of Astro AST representations earlier in the rendering process, eg when rendering a complete page.
- I've learnt that [portal](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/portal) is an existing, experimental but unused valid HTML element
- You said "Astro compiler remove unknown head elements" => isn't that a strange behaviour? Or is it expected to have valid HTML?

**Reply by Fryuni** [reply]
> A downside might be the cost, you need to parse the whole HTML content to find the portal. Ideally, this should be done against a kind of Astro AST representations earlier in the rendering process

Astro doesn't have such a representation. Components render directly to HTML by turning the code into a template with interpolation. You can see it happening [here][compiler example]. The template uses a `$$render` [tag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) that returns an async iterator over each part of the template string and the interpolated values.

Another thing present in that example is that Astro adds `${$$renderHead($$result)}` to the interpolation. That is where any component used can statically add elements to the head powered by the compiler. This is how `<script>` and `<style>` tags from all components in the page get added to the head. But since this happens before the components in the `<body>` are even called it only knows about what was discovered statically (this limitation is also why a component style will be present on the page even if you don't use it but import it, which has been reported multiple times).

It brings the performance of skipping any intermediate representation but, at the same time, limits what can be done in post-processing without parsing the HTML.

[compiler example]: https://live-astro-compiler.vercel.app/?editor-state=WzMxLDEzOSw4LDAsODQsMTkzLDI1MSwxMDIsMiwzLDExNyw4MiwyMDEsNzgsMTk1LDQ4LDE2LDI1MywxNDksMjAyLDE1MSw5NCwyMTgsMTMyLDIyOSw4NiwxNjIsNzQsMjEsODAsMTMyLDE5NiwxLDUzLDEwMSw1NywyMjgsMjI2LDE5OCwxOSwxMDYsMjAxLDI0Niw4OCw5OSwxMzUsMTc4LDEzNiwxMjcsMTAzLDE3OCw1MiwyMCw0MiwxMjQsMjQxLDI0Miw1NCwxNDMsMTk5LDE1OSwxNjIsMTcyLDEzNywxOTIsMTk3LDc1LDE4MCw5NCwyNywxNjAsNzEsMTYwLDE2MCwyMDksMTM3LDE1Myw1Niw3NSw3OCw3OSwxNDYsMTE1LDQ5LDE3LDM3LDQyLDIyNCwyNTMsMTE2LDU4LDQ1LDE1NiwxODIsMzAsNDEsMTQyLDE1MCwxMzYsMTYzLDEzOCwyMDgsMTQyLDE5OCw3MywyMDIsMjM1LDY4LDEzNCw3Miw1NiwxOTAsNDAsOTIsNzUsNDIsOTIsMTgyLDE0MSwyMTQsMjA0LDExLDU1LDIyNiwxNDUsMTA5LDY1LDE3MCwxMjYsMjIxLDIzOCwxNjMsMTQyLDYsMjMwLDg5LDIxOCwyMDUsNyw2NCwyMjcsMTU0LDIzOCw4NSwyMzMsMTI5LDQ0LDIxOSwxNjAsMTIyLDI1NSwxNTksMjE3LDE2Myw0NCwxMDUsODIsMjQ5LDE5OCw1OSwzNiwyNDUsNjgsMjEwLDEyMyw4MCw5OCw4Niw3MywxOSw5NiwzNCwxMDgsODcsNjksMzYsMjMzLDY2LDEzMywxMDAsMTUzLDIzMCwzNyw1LDE4NCwxOTksMTYwLDk5LDkxLDExMyw3OSwyOCwyNCwxODMsNDYsMiw1NywxMDUsMzAsODYsMTE5LDQ0LDEwMCwxMjYsMTk3LDE1LDIyOCwxNjQsMTAxLDI3LDg3LDI3LDUxLDE3LDE0Miw3MywyMTAsMjMyLDE1LDgwLDIwMyw2MywyMDAsMjI0LDE0NSw5OSw3NywzNyw4OCwyMzMsMTQzLDIzNywyMywyMDUsMTU1LDIyMSwyNCwyMjAsNzIsMTc5LDE2MCwxNTEsMjA4LDY5LDEyLDEwNCwyMTEsMTUsODksMTk4LDk5LDIxNywxMCw2NiwxMDksOTgsOTQsMzQsMjMsMTUxLDI3LDI1Miw5NywxODgsMTA2LDIxNiw5Myw2NSwxNDgsMTI0LDIxLDIxMywxNDUsMjE2LDE3OSwxOTksMiwyNCw0MCwzNSwxNjgsMjQ1LDIyMiwxMDEsNDUsNTUsMjgsMjE2LDExOCwyNDYsMCwyMDUsMTU5LDEyNywxNTcsMTUsMTUzLDIxNSwxMTEsMzAsNzIsOTEsMjU0LDM4LDIxMCwxNzIsMTkyLDQxLDE2MCwxODgsMzYsMjM3LDI0NywyMTcsOTUsMjIzLDkwLDExNSwxODYsMjMzLDcwLDIsMCwwXQ%3D%3D

> I've learnt that [portal](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/portal) is an existing, experimental but unused valid HTML element

Yes, I also found that out. Technically, any tag without a `-` is reserved by the HTML spec for future use. That is why custom elements for web components must have a `-`.

This is not really a problem here. Astro already overrides the semantics of `<slot>`, `<script>` and `<style>`. You have to opt-out of the Astro semantics back into the HTML semantics by adding `is:inline`. The same could be done for `<portal>` if we wanted.

> You said "Astro compiler remove unknown head elements" => isn't that a strange behaviour? Or is it expected to have valid HTML?

I think it is a bug. But it was 2 a.m., so I just figured out what was happening and not why it was happening.

It does [other weird things][weird] things as well. If you have a tag that is not supposed to be on the `<head>` inline there it doesn't emit the `<body>` tag

[weird]: https://live-astro-compiler.vercel.app/?editor-state=WzMxLDEzOSw4LDAsMTczLDE5NywyNTEsMTAyLDIsMywxMDEsODIsMjAxLDExMCwyLDQ5LDEyLDI1MywxNDksNDIsMTAzLDk2LDE4NiwyMjAsMjA4LDgsOSwxMTcsODMsMTY1LDMwLDQyLDEzNCw0Niw3LDQ2LDM4LDQ5LDM3LDE0NiwxNzksMjAwLDIwMSwxNDgsNDYsMjM0LDE5MSwyMTUsMjA0LDEyLDIxMSwxMDUsMjAxLDMzLDEzNywyNTMsMTU4LDE1OSwyMzcsNTYsOTUsNzQsMjE1LDIwNCwyMzIsMjQzLDEwMSwxMTIsMjA5LDE4LDI0MiwxOSwxMTQsMTc4LDE5MywxNzEsMTY5LDU4LDE1OSwxNTYsMTU3LDc4LDQ2LDIxMiw3MiwyMzMsOTYsODAsMjM2LDI0MSwxMjAsMTg4LDI0MiwyMDUsMTgyLDI0MiwyMjksNTQsNTksMTU0LDE3MywyNTIsMTM3LDE3MiwxMTQsMTM5LDk2LDE4NiwxMjMsOTksMTAzLDE1NSw5LDEwMywxMDEsMjA5LDE1OCwzLDMyLDYsMjA2LDY0LDEzMCwxMTYsMTUxLDQ2LDE5MCwyNCw4LDE0OCwyMzUsOTYsNjIsMTM0LDQ5LDE0LDE3MiwzMSwyMTgsMTk3LDE5Miw4MSwyMiwyOSw5MSwzNiwyNDYsMjQ1LDcyLDE3MywxODcsMTkyLDIzMCwxNTMsMzMsNzAsNTIsMTA2LDE4NiwxLDc0LDU2LDgyLDE3NCwxNzMsNjMsNTEsMjQ4LDE4MCw5LDIzNiwxMzIsMjIsMTI5LDE5LDYyLDEzMiwxMDAsMTE1LDIxMSwxMDcsNzEsMjM2LDI1LDExOSw2MiwzNSwxMjMsMTYwLDE5OSwxOTcsMTg5LDQsMTAsMTI3LDM1LDc5LDIyNywxOTMsMTM3LDE0MCwxNzUsMTM3LDcwLDIwMiwxMSw5LDIwMCwxMjYsMTYyLDE4NSwyNDksMTM1LDI0NCwyNiw4NSwxNjgsODksMTYzLDEzMSwxMjAsNDQsNjMsNzksMTUzLDE5NSw0NSwxMzMsNTMsMjA4LDE1Niw5NSw4MywxNTUsMTYyLDcxLDI0NywxNDcsMCwxNTcsMTQzLDE5NSwyMiwxNTIsMTA2LDIwMiwxNDksMTQsMjEwLDkyLDY5LDIyNSwxNTEsMjQxLDEwMiwxMTMsMTE5LDEzMywyNSwxNjQsMjAsMjExLDE0Niw2OCwxNzksMTk1LDE4LDE4LDIzNCwxNDAsMTAyLDEyMSw4MCw4OSwxOTQsOTAsMTgsNTQsNTEsMjksMTYwLDIxMywyMDMsMzEsMTI3LDE1OSwyNDMsMjUwLDYxLDM0LDkxLDM5LDMxLDQsMTA0LDEyOSwyMjIsMzIsODcsMTU0LDEwOSw2MCwyMjgsMjU0LDI1NCwxLDYyLDEwNywxMTMsMTM4LDY0LDIsMCwwXQ%3D%3D

**Reply by eric-burel** [reply]
> Another thing present in that example is that Astro adds `${$$renderHead($$result)}` to the interpolation.

Interesting, maybe we could figure a solution specific to the <head> element, because that's a significant use case. Many framework choose to have a specific solution for metadata, eg Next.js `metadata` export, packages like Helmet etc.

Thanks again for your detailed technical insights

**Comment by zemd** [comment]
Are there any new thoughts about this functionality? I was thinking about **not** adding React-portal functionality since Portals in React are pure client-side components and are related to the interactive content, but rather server-side global slots.

For instance, I would expect to have in the layout something like the following:

```
<slot global server:defer name="html:title" />
```

and inside any other component, I would like to use it as any other slot:

```
<title slot="html:title" />
```

Potentially, if the slot is placed at the end of the document, nothing can affect streaming, but I guess the `server:defer` thing also does not block streaming.

As far as I can see, there is no possible way to do this in the userland. However, I am pretty new to Astro.

**Reply by Fryuni** [reply]
`server:defer` is a directive for an Astro Component to be turned into a sub-request. It was something already doable on user-land that became much easier. The equivalent previous code would be to have the component under `src/pages` with `export const partial = true` and add a client JS to call it and insert the returned HTML in the page.

Anything without blocking streaming would only be doable on the client after the entire page has been streamed since they can affect parts of the document that came before, this would lead to flashing the original content.

Server-side portals is conceptually contradictory with streaming. Streaming is to send the information of earlier elements without them being affected by slower work on following components. Server-side portals is to have earlier elements affected by the result of following components, necessarily having to wait until _after_ they are done otherwise there wouldn't be a result.

Besides that part, what you said is entirely doable on userland (it is so doable that [I did it](https://inox-tools.fryuni.dev/portal-gun)). Just not using `slot` tags, since they are reserved. I don't think this feature would be good as yet another behavior of the `slot` tag, so even it were to be added to Astro core (which I'm not in favor or against) I would personally be against a syntax like that. Maybe overriding the behavior of the experimental `portal` tag and allow reverting to the default HTML behavior with `is:inline` like the other tags Astro overrides would be a better syntax.

**Reply by zemd** [reply]
the `server:defer` thing was my lack of understanding of this functionality, thanks for clarifying it. 

overall, my view on this was that the compiler would defer the slot calculation until the next event loop tick or compiler iteration, assuming it works asynchronously. that could potentially give more flexibility in how the code can be compiled and streamed. 

as a matter of taste, I still think extending slot functionality would look a little bit better than having another `portal` tag. but it is just my personal preference.
